### DISCUSSION TITLE
For SHA_256, I am not getting correct output.

### DISCUSSION THREAD

**Discussion by rdiwan2023** [openssl/openssl#23969] [Q&A] [answered]
Hi,
I am using below code to get the SHA_256 digest output, but I am seeing output is not correct. Can any one please tell am i missing any thing? below is the code.

#include <stdio.h>
#include <stdlib.h>
#include <openssl/err.h>
#include <openssl/bio.h>
#include <openssl/evp.h>
#include <openssl/core_names.h>

OSSL_LIB_CTX *libctx = NULL;
const char *propq = NULL;

int sha_256()
{
	int ret = 0;
	EVP_MD_CTX *ctx;
	const EVP_MD *md;

	//create md context
	ctx = EVP_MD_CTX_create();
	if(!ctx){
   	return -1;
	}

// initialise the MD to use sha256
    md = EVP_sha256();
    if (!md) {
        return -1;
    }

    // initialise the digest
    ret = EVP_DigestInit_ex(ctx, md, NULL);
    if (ret != 1) {
        return -1;
    }
    //std::string input = "hash function";
    //char str[] = "hash function";
    char str[] = "abc";
    int length = 0;
    length = sizeof(str);

    // update digest with the given input buffer
    ret = EVP_DigestUpdate(ctx, str, length);
   
    if (ret != 1) {
        return -1;
    }
    uint8_t sha256_hash[32];
    int sha256_hashlen = 0;

    // calculate the final hash
    ret = EVP_DigestFinal_ex(ctx, sha256_hash, (unsigned int *)&sha256_hashlen);
    if (ret != 1) {
        return -1;
    }
    
    // dump the final hash
    int i;
    printf("sha256: ");
    for (i = 0; i < sha256_hashlen; i ++) {
        printf("%0x", sha256_hash[i]);
    }
    printf("\n");

    // destroy md context
    EVP_MD_CTX_destroy(ctx);

 }   

int main(int argc, char **argv)
{
   printf("inside main function====\n");
   sha_256();
   return 0;
}

output : 
inside main function====
sha256: dc1114cd74914bd872cc1f9a23ec910ea223bc79779ab2e17da25782a624fc

Excpected output is:  ba7816bf8f01cfea414140de5dae2223b00361a396177a9cb410ff61f20015ad

Thanks
Rohith

**Comment by t8m** [comment]
The `length = sizeof(str);` means that you're including the terminating `'\0'` character in the digest of the string. Is this something you really wanted?

**Comment by rdiwan2023** [comment]
sorry, I didn't get what you're asking. I am passing the length of the array to  EVP_DigestUpdate(ctx, str, length);

**Reply by t8m** [reply]
Should you, in your code, use `length = sizeof(str) - 1;` or `length = strlen(str);` instead?

**Comment by nhorman** [comment, marked as answer]
@t8m is correct, sizeof(str) returns 4 (as it includes the NULL terminator on the char array, whereas strlen(str) would return 3.  If you run this code replacing sizeof(str) with strlen(str), the expected sha256 output is produced

**Reply by rdiwan2023** [reply]
Thanks @t8m & @nhorman , by making length = sizeof(str) - 1, expected digest is produced.
